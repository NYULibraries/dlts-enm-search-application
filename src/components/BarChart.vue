<template>
    <div id="bar-chart">
        BarChart
    </div>
</template>

<script>
export default {
    name: 'BarChart',
};
</script>

<style>
</style>
