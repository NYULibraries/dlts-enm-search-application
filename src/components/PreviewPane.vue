<template>
    <div
        v-show="display"
        class="column enm-pane enm-pane-preview">

        <!--PREVIEW: NO EPUB SELECTED-->
        <div
            v-show="! isbn"
            id="message"
            class="message is-valign is-centered">
            <p class="message-body has-text-centered">Choose a book at left to view matched pages, or
                <br>
                <a
                    href="#"
                    @click="loadFirstMatchedPage">Load the first matched page</a>
            </p>
        </div>
        <!--PREVIEW: NO EPUB SELECTED-->

        <!--PREVIEW-->
        <div>
            <span
                v-if="isbn"
                id="preview-isbn"
                :name="isbn"
                style="display: none"></span>

            <bar-chart
                :bar-chart-data="barChart.barChartData"
                :title="title"
                @bar-click="previewEpubPage"
            />

            <div
                v-show="selectedPageNumber"
                class="enm-topicsonthispage">
                <h3>Topics on this page</h3>

                <ul>

                    <li
                        v-for="topicOnPage in topicsOnPage"
                        :key="topicOnPage">
                        <a
                            href="#"
                            v-html="topicOnPage">
                        </a>
                    </li>
                </ul>

                <div
                    v-if="topicsOnPage === null || topicsOnPage.length === 0"
                    class="enm-notopics">
                    No topics are associated with this page.
                </div>

            </div>

            <hr>

            <div
                v-show="selectedPageNumber"
                class="enm-pageText"
                v-html="pageText">
            </div>
        </div>
    </div>
</template>

<script>
import BarChart from './BarChart';

import { mapGetters, mapActions } from 'vuex';

const ALTERNATE_NAMES_LIST_SEPARATOR = '&nbsp;&bull;&nbsp;';
const HIGHLIGHT_PRE = '<mark>';
const HIGHLIGHT_POST = '</mark>';

export default {
    name       : 'PreviewPane',
    components : { BarChart },
    props      : {
        currentSearch : {
            type     : Object,
            required : true,
            default  : null,
        },
        display       : {
            type     : Boolean,
            required : true,
            default  : false,
        },
        isbn          : {
            type     : String,
            required : true,
            default  : null,
        },
        title         : {
            type     : String,
            required : true,
            default  : null,
        },
    },
    data() {
        return {
            barChart           : {
                barChartData : [],
                isbn         : this.isbn,
                title        : this.title,
            },
            pageText           : null,
            selectedPageNumber : null,
            topicsOnPage       : null,
        };
    },
    computed : {
        ...mapGetters(
            [
                'query',
                'queryFields',
                'selectedTopicFacetItems',
            ]
        ),
    },
    watch: {
        isbn( newIsbn, oldIsbn ) {
            if ( newIsbn === '' ) {
                this.barChart.barChartData = [];
                this.pageText = null;
                this.selectedPageNumber = null;
                this.topicsOnPage = null;

                return;
            }

            this.previewEpub();
        },
    },
    methods: {
        ...mapActions(
            [
                'addSelectedTopicFacetItems',
                'removeSelectedTopicFacetItems',
                'setQuery',
                'setQueryFields',
            ]
        ),
        loadFirstMatchedPage() {
            this.$emit( 'load-first-matched-page-link-click' );
        },
        async previewEpub() {
            let response;
            try {
                response = await this.solrPreviewEpub(
                    this.isbn,
                    this.currentSearch.query,
                    this.currentSearch.queryFields,
                    this.currentSearch.selectedTopicFacetItems,
                );
            } catch( e ) {
                console.error( 'ERROR in PreviewPane.previewEpub: ' + e );

                // TODO: replace this with something more user-friendly
                alert( 'Sorry, the server has returned an error or is not responding.' );

                return;
            }

            const barChartData = [];

            // docs are sorted by pageSequenceNumber in asc order
            response.response.docs.forEach( ( doc ) => {
                barChartData.push(
                    {
                        page  : doc.pageNumberForDisplay,
                        score : doc.score,
                    }
                );
            } );

            this.barChart.barChartData = barChartData;
        },
        async previewEpubPage( pageNumberForDisplay ) {
            this.selectedPageNumber = pageNumberForDisplay;

            let response;
            try {
                response = await this.solrPreviewPage(
                    this.isbn,
                    this.selectedPageNumber,
                    this.currentSearch.query,
                    this.currentSearch.queryFields
                );
            } catch( e ) {
                console.error( 'ERROR in PreviewPane.previewEpubPage: ' + e );

                // TODO: replace this with something more user-friendly
                alert( 'Sorry, the server has returned an error or is not responding.' );

                return;
            }

            const doc = response.response.docs[ 0 ];
            const highlights = response.highlighting[
                Object.keys( response.highlighting )[ 0 ]
            ];

            let topicHighlights, topicsOnPage = [];
            let topicNamesLists;

            if ( highlights.pageText ) {
                this.pageText = highlights.pageText[ 0 ];
            } else {
                this.pageText = doc.pageText;
            }

            if ( highlights.topicNamesForDisplay ) {
                // Sample of prettified JSON string (without wrapping quotes):
                // [
                //     [
                //         "Conference on Critical Legal Studies"
                //     ],
                //     [
                //         "<mark>identity</mark> -- politics of",
                //         "<mark>Identity</mark> politics",
                //         "\"<mark>Identity</mark> politics\"",
                //         "Politics of <mark>identity</mark>"
                //     ],
                //     [
                //         "Ideology of the subject"
                //     ]
                // ]

                topicHighlights = JSON.parse( highlights.topicNamesForDisplay );
                topicHighlights.forEach( ( topicHighlightArray ) => {
                    const preferredName = topicHighlightArray.shift();
                    const alternateNames = topicHighlightArray;

                    if ( alternateNames.length === 0 ) {
                        // No alternate names
                        topicsOnPage.push( preferredName );
                    } else {
                        if ( namesListContainsHighlights( alternateNames ) ) {
                            // Display alternate names - they contain highlights
                            topicsOnPage.push(
                                preferredName +
                                ' <span class="enm-alt-names">(also: ' +
                                alternateNames.join( ALTERNATE_NAMES_LIST_SEPARATOR ) +
                                ')</span>'
                            );
                        } else {
                            // Do not display alternate names - they do not contain highlights
                            topicsOnPage.push( preferredName );
                        }
                    }
                } );

                this.topicsOnPage = topicsOnPage;
            } else if ( doc.topicNamesForDisplay ) {
                topicNamesLists = JSON.parse( doc.topicNamesForDisplay );
                this.topicsOnPage = topicNamesLists.map(
                    ( topicNamesList ) => {
                        // The display/preferred name is the first element
                        return topicNamesList.shift();
                    }
                );
            } else {
                this.topicsOnPage = [];
            }
        },
        async solrPreviewEpub( isbn, query, queryFields, selectedTopicFacetFields ) {
            const response = await this.$solrPreviewEpub(
                isbn,
                query,
                queryFields,
                selectedTopicFacetFields,
            );

            return response;
        },
        async solrPreviewPage( isbn, pageNumberForDisplay, query, queryFields ) {
            const response = await this.$solrPreviewPage(
                isbn,
                pageNumberForDisplay,
                query,
                queryFields
            );

            return response;
        },
    },
};

function namesListContainsHighlights( alternateNames ) {
    return alternateNames.filter( ( alternateName ) => {
        return alternateName.indexOf( HIGHLIGHT_PRE ) !== -1 &&
               alternateName.indexOf( HIGHLIGHT_POST ) !== -1;
    } ).length > 0;
}
</script>

<style>
</style>
