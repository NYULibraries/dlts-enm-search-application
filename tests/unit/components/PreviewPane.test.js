import BarChart from '@/components/BarChart';
import PreviewPane from '@/components/PreviewPane';

import merge from 'lodash.merge';
import { shallowMount } from '@vue/test-utils';

import { createLocalVueWithVuex, createStore } from '../test-utils';

const ISBN_BEFORE_BOOK_SELECTED  = '';
const TITLE_BEFORE_BOOK_SELECTED = '';

const ISBN                             = '9781111111111';
const QUERY                            = 'art';
const QUERY_FIELDS                     = [ 'pageText', 'topicNames' ];
const SELECTED_TOPIC_FIELD_FACET_ITEMS = [ 'art', 'drawing', 'painting' ];
const TITLE                            = 'The Book';

function createWrapper( overrides ) {
    const defaultMountingOptions = {
        propsData : {
            display : false,
            isbn    : ISBN_BEFORE_BOOK_SELECTED,
            title   : TITLE_BEFORE_BOOK_SELECTED,
        },
    };

    return shallowMount( PreviewPane, merge( defaultMountingOptions, overrides ) );
}

describe( 'PreviewPane', () => {
    describe( 'initialized with no book selected', () => {
        const LOAD_FIRST_MATCHED_PAGE_LINK_CLICK_EVENT = 'load-first-matched-page-link-click';

        let wrapper;

        beforeEach( () => {
            wrapper = createWrapper();
        } );

        test( `clicking on "Load the first matched page" emits "${ LOAD_FIRST_MATCHED_PAGE_LINK_CLICK_EVENT }" event`,
            () => {
                wrapper.find( '#load-first-matched-page-link' ).trigger( 'click' );

                expect( wrapper.emitted()[ LOAD_FIRST_MATCHED_PAGE_LINK_CLICK_EVENT ] ).toBeTruthy();
            }
        );

        test( 'renders correctly', () => {
            expect( wrapper.element ).toMatchSnapshot();
        } );
    } );

    describe( 'when a new book is selected', () => {
        const MOCK_SOLR_RESPONSE_DOCS         = [
            {
                'pageNumberForDisplay' : '5',
                'score'                : 2.6782167,
            },
            {
                'pageNumberForDisplay' : '19',
                'score'                : 2.6782167,
            },
            {
                'pageNumberForDisplay' : '126',
                'score'                : 4.766254,
            },
            {
                'pageNumberForDisplay' : '192',
                'score'                : 5.5347614,
            },
            {
                'pageNumberForDisplay' : '218',
                'score'                : 4.7891774,
            },
        ];
        const EXPECTED_BAR_CHART_DATA = MOCK_SOLR_RESPONSE_DOCS.map( doc => {
            return {
                page  : doc.pageNumberForDisplay,
                score : doc.score,
            };
        } );

        const localVue = createLocalVueWithVuex();
        const store = createStore( QUERY, QUERY_FIELDS, SELECTED_TOPIC_FIELD_FACET_ITEMS );

        let wrapper;

        beforeEach( () => {
            const $solrPreviewEpub = jest.fn().mockReturnValueOnce(
                {
                    'response' : {
                        'docs' : MOCK_SOLR_RESPONSE_DOCS,
                    },
                }
            );

            wrapper = createWrapper(
                {
                    mocks : {
                        $solrPreviewEpub,
                    },
                    store,
                    localVue,
                }
            );

            wrapper.setProps(
                {
                    display : true,
                    isbn    : ISBN,
                    title   : TITLE,
                }
            );
        } );

        test( 'calls $solrPreviewEpub with proper arguments', () => {
            expect( wrapper.vm.$solrPreviewEpub.mock.calls[ 0 ] ).toEqual(
                [
                    ISBN,
                    QUERY,
                    QUERY_FIELDS,
                    SELECTED_TOPIC_FIELD_FACET_ITEMS,
                ]
            );
        } );

        // We only need to test barChart.barChartData.
        // barChart.isbn and barChart.title are generally set one time only
        // through BarChart props assignment in template.
        test( 'passes BarChart component correct barChartData', () => {
            const barChartStub = wrapper.find( BarChart );
            expect( barChartStub.vm.barChartData ).toEqual( EXPECTED_BAR_CHART_DATA );
        } );
    } );

    describe( 'when BarChart emits a "bar-click" event', () => {
        const MOCK_SOLR_RESPONSE_PREVIEW_PAGE  =
                  require( '../fixtures/solr-responses/solr-preview-page-highlights-in-pagetext-and-topicnames' );
        const SELECTED_PAGE_NUMBER = 12;
        const localVue = createLocalVueWithVuex();
        const store = createStore( QUERY, QUERY_FIELDS, SELECTED_TOPIC_FIELD_FACET_ITEMS );

        let wrapper;

        beforeEach( () => {
            const $solrPreviewEpub = jest.fn();

            const $solrPreviewPage = jest.fn().mockReturnValueOnce(
                MOCK_SOLR_RESPONSE_PREVIEW_PAGE
            );

            wrapper = createWrapper(
                {
                    mocks : {
                        $solrPreviewEpub,
                        $solrPreviewPage,
                    },
                    store,
                    localVue,
                }
            );

            wrapper.setProps(
                {
                    display : true,
                    isbn    : ISBN,
                    title   : TITLE,
                }
            );

            wrapper.find( BarChart ).vm.$emit(
                'bar-click',
                SELECTED_PAGE_NUMBER
            );
        } );

        test( 'calls $solrPreviewPage with proper arguments', () => {
            expect( wrapper.vm.$solrPreviewPage.mock.calls[ 0 ] ).toEqual(
                [
                    ISBN,
                    SELECTED_PAGE_NUMBER,
                    QUERY,
                    QUERY_FIELDS,
                ]
            );
        } );

        test( 'renders correctly selected page', () => {
            expect( wrapper.element ).toMatchSnapshot();
        } );
    } );
} );
