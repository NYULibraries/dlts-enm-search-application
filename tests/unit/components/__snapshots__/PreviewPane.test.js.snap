// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PreviewPane initialized with no book selected renders correctly 1`] = `
<div
  class="column enm-pane enm-pane-preview"
  style="display: none;"
>
  <div
    class="message is-valign is-centered"
    id="message"
  >
    <p
      class="message-body has-text-centered"
    >
      
            Choose a book at left to view matched pages, or
            
      <br />
       
      <a
        href="#"
        id="load-first-matched-page-link"
      >
        
                Load the first matched page
            
      </a>
    </p>
  </div>
   
  <div>
    <!---->
     
    <barchart-stub
      barchartdata=""
      title=""
    />
     
    <h2
      class="is-size-4"
      id="preview-epub-error"
      style="display: none;"
    >
      
            Sorry, this book cannot be previewed due to a server error.
            Please try your search again later.
        
    </h2>
     
    <div
      class="enm-topicsonthispage"
      style="display: none;"
    >
      <h3>
        Topics on this page
      </h3>
       
      <ul />
       
      <div
        class="enm-notopics"
      >
        
                No topics are associated with this page.
            
      </div>
    </div>
     
    <hr />
     
    <h2
      class="is-size-4"
      id="preview-epub-page-error"
      style="display: none;"
    >
      
            Sorry, these pages cannot be previewed due to a server error.
            Please try your search again later.
        
    </h2>
     
    <div
      class="enm-pageText"
      style="display: none;"
    />
  </div>
</div>
`;

exports[`PreviewPane when BarChart emits a "bar-click" event for a page hit with highlights in both page text and topic names renders preview correctly 1`] = `
<div
  class="column enm-pane enm-pane-preview"
  style=""
>
  <div
    class="message is-valign is-centered"
    id="message"
    style="display: none;"
  >
    <p
      class="message-body has-text-centered"
    >
      
            Choose a book at left to view matched pages, or
            
      <br />
       
      <a
        href="#"
        id="load-first-matched-page-link"
      >
        
                Load the first matched page
            
      </a>
    </p>
  </div>
   
  <div>
    <span
      id="preview-isbn"
      name="9781111111111"
      style="display: none;"
    />
     
    <barchart-stub
      barchartdata=""
      title="The Book"
    />
     
    <h2
      class="is-size-4"
      id="preview-epub-error"
      style="display: none;"
    >
      
            Sorry, this book cannot be previewed due to a server error.
            Please try your search again later.
        
    </h2>
     
    <div
      class="enm-topicsonthispage"
      style=""
    >
      <h3>
        Topics on this page
      </h3>
       
      <ul>
        <li>
          <a
            href="#"
          >
            <mark>
              Lorem
            </mark>
             
            <mark>
              ipsum
            </mark>
             dolor 
            <span
              class="enm-alt-names"
            >
              (also: dolor • L.I. • L.I.D. • LID • 
              <mark>
                Lorem
              </mark>
               
              <mark>
                Ipsum
              </mark>
               • 
              <mark>
                Lorem
              </mark>
               
              <mark>
                Ipsum
              </mark>
               (L.I.) • 
              <mark>
                Lorem
              </mark>
               
              <mark>
                Ipsum
              </mark>
               (LI))
            </span>
          </a>
        </li>
        <li>
          <a
            href="#"
          >
            vitae
          </a>
        </li>
        <li>
          <a
            href="#"
          >
            maximus
          </a>
        </li>
        <li>
          <a
            href="#"
          >
            felis
          </a>
        </li>
      </ul>
       
      <!---->
    </div>
     
    <hr />
     
    <h2
      class="is-size-4"
      id="preview-epub-page-error"
      style="display: none;"
    >
      
            Sorry, these pages cannot be previewed due to a server error.
            Please try your search again later.
        
    </h2>
     
    <div
      class="enm-pageText"
      style=""
    >
      <mark>
        Lorem
      </mark>
       
      <mark>
        ipsum
      </mark>
       dolor sit amet, consectetur adipiscing elit. Curabitur pretium 
      <mark>
        lorem
      </mark>
       elit, in lobortis erat posuere at. Ut a nulla iaculis, tristique tortor at, consectetur arcu. Donec consectetur purus ligula, ac congue diam eleifend at. Nulla semper 
      <mark>
        lorem
      </mark>
       sit amet quam vulputate pharetra. Quisque a tincidunt urna. Quisque dapibus condimentum felis, lobortis sagittis purus hendrerit vitae. Mauris augue libero, blandit ac orci quis, congue porta nibh. Phasellus nec tellus auctor, tincidunt eros vitae, maximus arcu.

Integer eu euismod sapien, vitae condimentum risus. Donec ut porttitor nunc, at sodales ante. Curabitur felis nisl, aliquet vel 
      <mark>
        lorem
      </mark>
       in, faucibus dignissim nibh. Cras lacinia pharetra purus a tempus. Fusce quam velit, imperdiet quis velit vel, dignissim hendrerit arcu. Cras iaculis consequat urna ut vehicula. Nunc et leo eget tellus faucibus rhoncus nec ac velit. Cras scelerisque elementum ex non volutpat. Ut aliquet lectus et neque fermentum tempor.

Phasellus sed orci ac arcu lacinia volutpat. Proin semper pharetra turpis in rhoncus. Fusce nulla leo, ornare quis lobortis ac, ornare at ex. Donec gravida ut sem ut elementum. Vivamus porttitor nibh sed accumsan mollis. Nunc eu blandit felis. Sed laoreet lectus id nisl tempus consequat. Praesent ullamcorper porta mollis.

Nulla facilisis lectus vitae magna molestie, sit amet interdum sem pretium. Fusce pharetra aliquam metus in finibus. Maecenas efficitur dapibus dui non posuere. Fusce tristique turpis arcu, vitae congue massa ornare in. Etiam et luctus nisi. Integer sem magna, fringilla at dui consequat, congue condimentum purus. Aliquam sit amet vulputate nunc. Proin accumsan diam vel consequat vehicula. Quisque magna odio, vestibulum molestie metus non, sollicitudin semper quam.

Sed et sodales orci, sed laoreet eros. Vestibulum viverra lacus accumsan turpis tristique convallis. 
      <mark>
        Lorem
      </mark>
       
      <mark>
        ipsum
      </mark>
       dolor sit amet, consectetur adipiscing elit. 
      <mark>
        Lorem
      </mark>
       
      <mark>
        ipsum
      </mark>
       dolor sit amet, consectetur adipiscing elit. Sed vitae faucibus neque. Morbi blandit felis quis fringilla auctor. Fusce mi erat, imperdiet pharetra elit vel, pulvinar facilisis elit. Nulla aliquam magna sed ligula mattis tempor. Quisque rutrum est sit amet 
      <mark>
        ipsum
      </mark>
       posuere, auctor facilisis nisi accumsan. Proin ultricies justo in tortor pharetra consequat. Suspendisse potenti. Praesent interdum nunc sed enim gravida porta.
    </div>
  </div>
</div>
`;

exports[`PreviewPane when BarChart emits a "bar-click" event for a page hit with highlights in page text only renders preview correctly 1`] = `
<div
  class="column enm-pane enm-pane-preview"
  style=""
>
  <div
    class="message is-valign is-centered"
    id="message"
    style="display: none;"
  >
    <p
      class="message-body has-text-centered"
    >
      
            Choose a book at left to view matched pages, or
            
      <br />
       
      <a
        href="#"
        id="load-first-matched-page-link"
      >
        
                Load the first matched page
            
      </a>
    </p>
  </div>
   
  <div>
    <span
      id="preview-isbn"
      name="9781111111111"
      style="display: none;"
    />
     
    <barchart-stub
      barchartdata=""
      title="The Book"
    />
     
    <h2
      class="is-size-4"
      id="preview-epub-error"
      style="display: none;"
    >
      
            Sorry, this book cannot be previewed due to a server error.
            Please try your search again later.
        
    </h2>
     
    <div
      class="enm-topicsonthispage"
      style=""
    >
      <h3>
        Topics on this page
      </h3>
       
      <ul>
        <li>
          <a
            href="#"
          >
            Lorem ipsum dolor
          </a>
        </li>
        <li>
          <a
            href="#"
          >
            vitae
          </a>
        </li>
        <li>
          <a
            href="#"
          >
            maximus
          </a>
        </li>
        <li>
          <a
            href="#"
          >
            felis
          </a>
        </li>
      </ul>
       
      <!---->
    </div>
     
    <hr />
     
    <h2
      class="is-size-4"
      id="preview-epub-page-error"
      style="display: none;"
    >
      
            Sorry, these pages cannot be previewed due to a server error.
            Please try your search again later.
        
    </h2>
     
    <div
      class="enm-pageText"
      style=""
    >
      <mark>
        Lorem
      </mark>
       
      <mark>
        ipsum
      </mark>
       dolor sit amet, consectetur adipiscing elit. Curabitur pretium 
      <mark>
        lorem
      </mark>
       elit, in lobortis erat posuere at. Ut a nulla iaculis, tristique tortor at, consectetur arcu. Donec consectetur purus ligula, ac congue diam eleifend at. Nulla semper 
      <mark>
        lorem
      </mark>
       sit amet quam vulputate pharetra. Quisque a tincidunt urna. Quisque dapibus condimentum felis, lobortis sagittis purus hendrerit vitae. Mauris augue libero, blandit ac orci quis, congue porta nibh. Phasellus nec tellus auctor, tincidunt eros vitae, maximus arcu.

Integer eu euismod sapien, vitae condimentum risus. Donec ut porttitor nunc, at sodales ante. Curabitur felis nisl, aliquet vel 
      <mark>
        lorem
      </mark>
       in, faucibus dignissim nibh. Cras lacinia pharetra purus a tempus. Fusce quam velit, imperdiet quis velit vel, dignissim hendrerit arcu. Cras iaculis consequat urna ut vehicula. Nunc et leo eget tellus faucibus rhoncus nec ac velit. Cras scelerisque elementum ex non volutpat. Ut aliquet lectus et neque fermentum tempor.

Phasellus sed orci ac arcu lacinia volutpat. Proin semper pharetra turpis in rhoncus. Fusce nulla leo, ornare quis lobortis ac, ornare at ex. Donec gravida ut sem ut elementum. Vivamus porttitor nibh sed accumsan mollis. Nunc eu blandit felis. Sed laoreet lectus id nisl tempus consequat. Praesent ullamcorper porta mollis.

Nulla facilisis lectus vitae magna molestie, sit amet interdum sem pretium. Fusce pharetra aliquam metus in finibus. Maecenas efficitur dapibus dui non posuere. Fusce tristique turpis arcu, vitae congue massa ornare in. Etiam et luctus nisi. Integer sem magna, fringilla at dui consequat, congue condimentum purus. Aliquam sit amet vulputate nunc. Proin accumsan diam vel consequat vehicula. Quisque magna odio, vestibulum molestie metus non, sollicitudin semper quam.

Sed et sodales orci, sed laoreet eros. Vestibulum viverra lacus accumsan turpis tristique convallis. 
      <mark>
        Lorem
      </mark>
       
      <mark>
        ipsum
      </mark>
       dolor sit amet, consectetur adipiscing elit. 
      <mark>
        Lorem
      </mark>
       
      <mark>
        ipsum
      </mark>
       dolor sit amet, consectetur adipiscing elit. Sed vitae faucibus neque. Morbi blandit felis quis fringilla auctor. Fusce mi erat, imperdiet pharetra elit vel, pulvinar facilisis elit. Nulla aliquam magna sed ligula mattis tempor. Quisque rutrum est sit amet 
      <mark>
        ipsum
      </mark>
       posuere, auctor facilisis nisi accumsan. Proin ultricies justo in tortor pharetra consequat. Suspendisse potenti. Praesent interdum nunc sed enim gravida porta.
    </div>
  </div>
</div>
`;
